Add retry tests for request bodies

dagnir opened the pull request:
This commit adds tests for implementations of `RequestBody` and `AsyncRequestBody` where the requests are retried to ensure that they send the same content for every attempt.

<!--- Provide a general summary of your changes in the Title above -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## Modifications
<!--- Describe your changes in detail -->

## Testing
<!--- Please describe in detail how you tested your changes -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License
<!--- The SDK is released under the Apache 2.0 license (http://aws.amazon.com/apache2.0/), so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a Contributor License Agreement (http://en.wikipedia.org/wiki/Contributor_License_Agreement) -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [ ] I confirm that this pull request can be released under the Apache 2 license

dagnir committed:
Add retry tests for request bodies

This commit adds tests for implementations of `RequestBody` and
`AsyncRequestBody` where the requests are retried to ensure that they send the
same content for every attempt.

dagnir committed:
Merge branch 'master' into dongie/streaming-tests

Fred1155 commented on test/s3-tests/src/test/java/software/amazon/awssdk/services/s3/BaseRequestBodyRetryTest.java:
```diff
@@ -0,0 +1,228 @@
+        return "aws-chunked".equalsIgnoreCase(request.getHeader("content-encoding"));
+    }
+
+    protected enum BodySize {
```
Do we need to add cases for 0 MB body size?

dagnir replied:
Good idea, will add one

dagnir replied:
Updated

dagnir committed:
Add 0 byte test

dagnir committed:
Merge branch 'master' into dongie/streaming-tests

dagnir committed:
Merge branch 'master' into dongie/streaming-tests

dagnir committed:
Merge branch 'master' into dongie/streaming-tests

dagnir committed:
Try to reduce memory usage

dagnir committed:
Switch to plain Jetty server

Unfortunately WireMock saves all the request bodies which increases
memory usage. Use a plain Jetty server that doesn't save requests to
avoid this. We don't rely on any WireWock request assertions so it works
out.

dagnir committed:
Merge branch 'master' into dongie/streaming-tests

dagnir committed:
Lower max size

dagnir committed:
Max 8MB

dagnir committed:
Force s3-tests to run at end

dagnir committed:
Merge branch 'master' into dongie/streaming-tests

dagnir committed:
Merge branch 'master' into dongie/streaming-tests

dagnir committed:
Increase memory

dagnir committed:
Merge branch 'master' into dongie/streaming-tests
